synapse_spamcheck_badlist/bad_list_filter.py

```diff
@@ -15,107 +15,108 @@
+
+        # A ahocorasick.Automaton used to recognize bad links.
+        self._link_automaton = None
+
+        # Start the loop to update links.
+        self._update_links_loop = LoopingCall(
+            lambda: defer.ensureDeferred(self._update_links_automaton())
+        )
+        self._update_links_loop.start(pull_from_db_every_sec)
+        # As soon as we can, run the first fetch.
+        # Note that we have no guarantee that this is finished
+        # by the time we receive the first message, so we need
+        # a fallback in `_get_links_automaton()`.
+        reactor.callWhenRunning(
+            lambda: defer.ensureDeferred(self._update_links_automaton())
         )
-        self._scheme_re = re.compile("https?:/*|git:/*|ftp:/*")
-        self._linkify_test_performance = Histogram('linkify_test_performance', 'Performance of calls to linkifier.test, in seconds')
-        self._link_test_performance = Histogram('link_test_performance', 'Total performance cost of checking links in a message, in seconds')
-
-        # One of:
-        # - `None` if we haven't checked yet whether the database is present;
-        # - `True` if we have checked and the database is present;
-        # - `False` if we have checked and the database is absent.
-        self._can_we_check_links = None
-        self._can_we_check_md5 = None
-
-    async def can_we_check_links(self) -> bool:
+
+    async def _update_links_automaton(self):
         """
-            Check whether the links database exists, caching the result.
+        Fetch the latest version of the links from the table, build an automaton.
         """
-        if self._can_we_check_links is not None:
-            return self._can_we_check_links
-        if self._links_table is None:
-            self._can_we_check_links = False
-            return False
-        try:
-            def interaction(db):
-                db.execute("SELECT url FROM %s LIMIT 1" % self._links_table)
-            await self._api.run_db_interaction("Check whether we can check links", interaction)
-            self._can_we_check_links = True
-            logger.info("We can check links!")
-        except Exception as e:
-            logger.warn("We CANNOT check links! %s" % e)
-            self._can_we_check_links = False
-        return self._can_we_check_links
-        
-    async def can_we_check_md5(self) -> bool:
+        logger.info(
+            "_update_links_automaton: fetching links from table %s" % self._links_table
+        )
+        links = await self._api.run_db_interaction(
+            "Fetch links from the table", _db_fetch_links, self._links_table
+        )
+        logger.info("_update_links_automaton: we received %s links" % len(links))
+        self._link_automaton = Automaton(ahocorasick.STORE_LENGTH)
+        for link in links:
+            self._link_automaton.add_word(link)
+        self._link_automaton.make_automaton()
```

Comment:
Actually, let's run this in a thread pool, as this might be an expensive call (especially if we start seeing more links)

```suggestion
        threads.deferToThread(self._link_automaton.make_automaton)
```

c.f. https://twistedmatrix.com/documents/current/api/twisted.internet.threads.html#deferToThread